File: 4-median-of-two-sorted-arrays/4-median-of-two-sorted-arrays.cs
public class Solution
{
    public double FindMedianSortedArrays(int[] nums1, int[] nums2)
    {
        if (nums1.Length > nums2.Length)
        {
            return FindMedianSortedArrays(nums2, nums1);
        }
        
        int m = nums1.Length;
        int n = nums2.Length;
        int mid = (m + n) / 2;
        
        int lo = 0;
        int hi = m;
        
        while (lo <= hi)
        {
            int cut1 = (lo + hi) / 2;
            int cut2 = mid - cut1;
            
            int L1 = cut1 == 0 ? int.MinValue : nums1[cut1 - 1];
            int L2 = cut2 == 0 ? int.MinValue : nums2[cut2 - 1];
            
            int R1 = cut1 == m ? int.MaxValue : nums1[cut1];
            int R2 = cut2 == n ? int.MaxValue : nums2[cut2];
            
            if (L1 > R2) 
            {
                hi = cut1 - 1;
            } 
            else if (L2 > R1) 
            {
                lo = cut1 + 1;
            }
            else
            {
                return (
                    (m + n) % 2 == 0
                    ? (double)(Math.Max(L1, L2) + Math.Min(R1, R2)) / 2
                    : Math.Min(R1, R2));
            }
        }
        
        return -1.0;
    }
}

/*
  while (lo <= hi) {
    if (L1 > R2) {
      hi = cut1 - 1
    } else if (L2 > R1) {
      lo = cut1 + 1
    } else {
      return ((m + n) % 2 === 0)
        ? (Math.max(L1, L2) + Math.min(R1, R2)) / 2
        : Math.min(R1, R2)
    }
  }
  
  return -1;
*/

File: 4-median-of-two-sorted-arrays/4-median-of-two-sorted-arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function(nums1, nums2) {
  if (nums1.length > nums2.length) {
    return findMedianSortedArrays(nums2, nums1);
  }
  
  var m = nums1.length;
  var n = nums2.length;
  
  var mid = (m + n) >> 1; // integer divide by two
  
  var lo = 0;
  var hi = m;
  
  var cut1, cut2, L1, L2, R1, R2;
  
  while (lo <= hi) {
    cut1 = (lo + hi) >> 1;
    cut2 = mid - cut1;
    
    L1 = (cut1 == 0) ? -Infinity : nums1[cut1 - 1];
    L2 = (cut2 == 0) ? -Infinity : nums2[cut2 - 1];
    
    R1 = (cut1 == m) ? Infinity : nums1[cut1];
    R2 = (cut2 == n) ? Infinity : nums2[cut2];
    
    if (L1 > R2) {
      hi = cut1 - 1;
    } else if (L2 > R1) {
      lo = cut1 + 1
    } else {
      return ((m + n) % 2 === 0)
        ? (Math.max(L1, L2) + Math.min(R1, R2)) / 2
        : Math.min(R1, R2)
    }
  }
  
  return -1
};
/*
      # Binary search for the correct cut index.
      if L1 > R2:
          hi = cut1 - 1
      elif L2 > R1:
          lo = cut1 + 1
      else:
          if (m + n) % 2:
              return min(R1, R2)
          return (max(L1, L2) + min(R1, R2)) / 2
  */